### ISSUE TITLE
Syntax for continuing by relative position

### ISSUE THREAD

**Issue by scy** [scy/timesheet.txt#13]
I'm splitting #3 in three separate issues, this is one of them. Example file:

```
1234  some entry
1255  customer calls
1315. # bathroom break
```

It's now 1322 and I want to continue a previous entry. Continuing `customer calls` is easy, we already have a syntax for that: `1322^`. But we want to continue `some entry`.

My initial suggestion was to use a Git-like syntax: `^` means the previous entry, `^^` means the one before that and `^^^` the one before __that__ one. So, in our example, we'd write `1322^^`.

I had also suggested to support, like Git, the syntax `^123`, meaning "123 entries ago". But as @rohieb pointed out in https://github.com/scy/timesheet.txt/issues/3#issuecomment-502745752, for more than three `^` characters you'd have to start counting lines backwards, which is complicated and error-prone. Therefore (and after discussing this a bit more with @L3viathan [on Twitter](https://twitter.com/scy/status/1140658567990693889)) I'd now say that the `^n` syntax should not be supported at all. I won't artificially limit the number of consecutive caret characters (insert CCC joke here) people are allowed to use though. Let them get messy if they really want to. 🤷‍♂️

Also, `^` is not referencing one _line_ above, but one _entry start_ above, of course. This means that in

```
1234  some entry
1255  customer calls
1315. # bathroom break
1322^^
```

the last line will naturally _not_ point to the `1315` "stop" line, but to the `1234` line that starts an entry. Everything else wouldn't make sense, you can't continue a "stop".

Continuation lines count as referenceable lines though, i.e.

```
1140  tidy up downloads folder
1234  some entry
1255  customer calls
1315. # bathroom break
1322^^
1350. # fetch a coffee
1355^^^^
```

that last line would continue the entry from `1140`. Basically, a continuation line should be handled as if the original entry was copy-pasted there (just with a different timestamp).

And indeed it starts to get confusing at that point. The best practices (see #9) should definitely recommend against doing relative back-references across more than three lines.

**Comment by scy** [comment]
One more thought:

> Continuation lines count as referenceable lines though

But should they really? If you have a file like this

```
1117  whatever
1234  foo
1310.
1320^
1340.
1355^
1420.
1425^
[…]
```

each continuation line after `1320` does not reference the original entry anymore, but the continuation line that came before it. In the code, you could of course store a pointer to the ultimate destination on each continuation line and thus prevent the creation of a long linked list, but what about the user writing that file?

For example, if I want to continue the `1117` line, I'd have to write something like `1505^^^^^` (or reference it by timestamp or something, of course).

If instead we defined `^` to only look for lines that are _not_ back-referencing another line, we could keep all of the `1320^`/`1355^`/`1425^` lines as they are (they would all directly refer to `1234` instead of possibly their previous continuation line), but to refer to `1117` I could now simply write `1505^^`.

**But!** 🍑 Let's suppose our file looks like this:

```
2019-11-15:
0815  work on christmas present
1230.

# [500 more lines]

2019-12-21:
0920  extensive breakfast
1510^?christmas  # using the syntax to refer to the 2019-11-15 item by string content
1620.
1655^  # would now _not_ continue the 1510 entry, but the 0920 one
```

In this example, continuing an entry that has been originally started quite a while ago, but continued recently, isn't so easy anymore.

A solution to that would be to say that `^` does not ignore _all_ continuation lines, but only `^` ones, but I feel that's inconsistent. Opinions?

**Comment by L3viathan** [comment]
What if we squash identical adjacent entries? Then you could use `1505^^` to refer to the `1117` line. The last example would also work as expected: `1655^` would now continue the `1510` entry.

Is this unintuitive?

**Comment by scy** [comment]
> Is this unintuitive?

To me it is, yeah. 😐 Also, again it would require more intelligence from an editor plugin.

Currently, I'm leaning towards allowing the inconsistency and defining `^` as finding a line that starts an entry, but ignoring `^` (or `^^`, or `^^^`) lines. In other words, it matches non-reference lines, but also `^1234` and `^?substring` type back-reference lines. This is reasonably simple to implement, also in an editor plugin, and actually feels rather intuitive to me.

Are there any good reasons against going this way?

**Comment by Zegnat** [comment]
> […] defining `^` as finding a line that starts an entry, but ignoring `^` (or `^^`, or `^^^`) lines.

I am having a bit of a hard time wrapping my head around how this would work for me. I have been giving this timesheet format a try and I often find myself doing things like: (Comments added for clarity.)

```
1100  Working on my issue
1130  Lunch
1200^^ # referring back to pre lunch activity
1215  Consult on coworker issue
1230^^ # referring back to previous backref
1500  Coffee break
1515^^ # again, previous backref
```

Basically `^^` is my biggest friend in this format. With the changes discussed here I would have to start remembering my original start time (to do `^1100`) or some other context (for a substring match). But when I get back to my desk, back to the issue at hand, I really just want to log the fact that I am continuing with my previous task without needing to remember when & how I originally tagged it.

If the relative position backrefs no longer refer to themselves, the above file will look something like this:

```
1100  Working on my issue
1130  Lunch
1200^^ # two steps to my issue
1215  Consult on coworker issue
1230^^^ # now it is three steps
1500  Coffee break
1515^^^^ # now it is four steps
```

Rather than having to remember the original start time, or some matchable string context, I now need to remember the amount of interruptions? This may be interesting to track as a metric in itself, but it definitely does not help me quickly add a line to the log.

TLDR: `^^` is an important way for me to log that I am continuing whatever I was doing before some interruption. I’d love to keep it part of the format.

**Comment by scy** [comment]
Thanks @Zegnat, that's very valuable feedback and exactly what I was asking for.  I agree that your `^^` use case is a valid one and should be easily accessible for a user.

The obvious solution to that would be to define `^` as looking for any timestamped line that is _not_ a stop line, i.e. match `1230 asdf`, `1230^^^`, `1230^1120` and so on. However, this inevitably leads to the problem I've talked about at the start of https://github.com/scy/timesheet.txt/issues/13#issuecomment-503156663. Got any ideas for that? Or should we accept

> or reference it by timestamp or something, of course

as the solution to it? I'd be okay with that, but if anyone has a better suggestion, let me know.

**Comment by dd23** [comment]
I though about it for a bit and think there are two options. I think this post just sums up what has already been proposed, just explained a bit differently, I guess?


**Option 1**, which I guess is rather simple and more straight-forward to implement solution would be for `^` to simply match everything that is not a stop line, i.e. simply pointy `^`-many (non-stop) lines up.

That would mean, that `1505^^` matches the entry at 1355, which points to 1320 and thus eventually point to `1234 foo`.
```
1117  whatever
1234  foo
1310.
1320^
1340.
1355^
1420.
1425^
1505^^
```

In your [example](https://github.com/scy/timesheet.txt/issues/13#issuecomment-503156663), where you want to refer from 1505 to 1117, I would use `1505^1117`, since counting more than, say, 3 or 4 `^` symbols is a pain.


**Option 2**, which is a bit more involved to implement but maybe more logical would be for `^` to match *logical entries* that happen before.
In the example the entries 1234, 1320, 1355, and 1425 together would be counted as a single entry. I think @L3viathan called this squashing.
That would mean, that `1505^^` in the example above matches the second logical entry above, which is `1117 whatever`.

Problem with Option 2 (added `1410 bar`):
```
1117  whatever
1234  foo
1310.
1320^
1340.
1355^
1410 bar
1420.
1425^^ #points to 1234 foo
1505^^^
```

The question that now arises is: Does `1505^^^` mean that we point to `1234 foo` or `1117 whatever`?, or put differently, what is a single entry? Does it have to be continuous (ignoring stop lines), or can it be interrupted by other events (`1410 bar`)? I think I would count *different* events, i.e. allow events to be interrupted and have `1505^^^` refer to `1117 whatever`.
Does that make sense?

In general, I think that stop lines should never match `^`.

**Comment by Zegnat** [comment]
> However, this inevitably leads to the problem I've talked about at the start of https://github.com/scy/timesheet.txt/issues/13#issuecomment-503156663. Got any ideas for that?

I agree that having to do `1505^^^^^` to skip past a long list of what is basically the same task may be a problem too. It just isn’t a problem I have encountered myself. And I dislike optimising for theoretical problems. Whenever I have found myself going back to something I worked on longer than 2 items or a couple hours ago, I am way more like to either:

1. Just write a completely new log line for it because I do not feel like finding the old one again. I can always use ticket IDs to deduplicate later.
2. Check back in the log when I worked on it before. At this point I will find the exact timestamp so writing `1505^1117` is now the easiest thing to do.

I like to think of the references as pointing to “the thing I did before I did this other thing”. Here the “squashing” does make sense. Probably. But again, not something that really has come up for me.
